File: src/lib/assembler.ts
import { dec2bin } from "./helper";

const instructionSet: { [code: string]: number } = {
    "HLT": 0b1111,
    "ADD": 0b1110,
    "SUB": 0b1101,
    "NOR": 0b1100,
    "AND": 0b1011,
    "XOR": 0b1010,
    "MOV": 0b1001,
    "LDR": 0b1000,
    "ADR": 0b0111,
    "JMP": 0b0110,
    "BRH": 0b0101,
    "CAL": 0b0100,
    "RET": 0b0011,
    "PGE": 0b0010,
    "LOD": 0b0001,
    "STR": 0b0000
}

const registerSet: { [code: string]: number } = {
    "r0":  0b0000,
    "r1":  0b0001,
    "r2":  0b0010,
    "r3":  0b0011,
    "r4":  0b0100,
    "r5":  0b0101,
    "r6":  0b0110,
    "r7":  0b0111,
    "r8":  0b1000,
    "r9":  0b1001,
    "r10": 0b1010,
    "r11": 0b1011,
    "r12": 0b1100,
    "r13": 0b1101,
    "r14": 0b1110,
    "r15": 0b1111
}

export class Assembler {
    parse(code: string): string[][] {
        const lines = code.split("\n");
        const parsed: string[][] = [];

        for (const line of lines) {
            const trimmed = line.trim();
            if (trimmed.length === 0 || trimmed.startsWith(";")) {
                continue
            };
            const parts = trimmed.split(" ");
            parsed.push(parts);
        };

        return parsed
    }

    assemble(parsed: string[][]): Uint16Array {
        let assembled: Uint16Array = new Uint16Array([]);
        let labels: { [label: string]: number } = {};
        let defintions: { [label: string]: number } = {};
        let instructions: number[][] = [];
        let address = 0;

        // pass one, translate defintions
        for (const parts of parsed) {
            if (parts.length === 3 && parts[0] === "define") {
                defintions[parts[1]] = parseInt(parts[2]);
            }
        }

        // pass two, translate labels
        for (const parts of parsed) {
            if (parts.length === 1 && parts[0].endsWith(":")) {
                const label = parts[0].slice(0,-1);
                labels[label] = address;
            } else {
                address++;
            }
        }

        // pass four, translate instructions
        for (let pc = 0; pc < parsed.length; pc++) {
            const parts = parsed[pc];
            let instruction: number[] = [];

            if (parts.length === 1 && parts[0].endsWith(":")) {
                continue;
            }
            if (parts[0] === "define") {
                continue;
            }

            const opcode = instructionSet[parts[0]];

            if (opcode === undefined) {
                throw new Error(`Unknown opcode: ${parts[0]}`);
            } else {
                instruction.push(opcode);
            }

            // check for correct operands length
            if (["HLT","RET"].includes(parts[0]) && parts.length !== 1) {
                throw new Error(`Invalid number of operands for instruction: ${parts[0]}`);
            }
            if (["JMP","CAL","PGE"].includes(parts[0]) && parts.length !== 2) {
                throw new Error(`Invalid number of operands for instruction: ${parts[0]}`);
            }
            if (["LOD","STR","LDR","ADR","BRH","MOV"].includes(parts[0]) && parts.length !== 3) {
                throw new Error(`Invalid number of operands for instruction: ${parts[0]}`);
            }
            if (["ADD","SUB","NOR","AND","XOR"].includes(parts[0]) && parts.length !== 4) {
                throw new Error(`Invalid number of operands for instruction: ${parts[0]}`);
            }

            // check for correct operands
            // REG A
            if (["ADD","SUB","NOR","AND","XOR","MOV","LDR","ADR","PGE","LOD","STR"].includes(parts[0])) {
                if (registerSet[parts[1]] === undefined) {
                    throw new Error(`Invalid reg A for ${parts[0]} on line ${pc + 1}`);
                }
                instruction.push(registerSet[parts[1]]);
            }

            // REG B
            if (["ADD","SUB","NOR","AND","XOR","MOV"].includes(parts[0])) {
                if (registerSet[parts[2]] === undefined) {
                    throw new Error(`Invalid reg B for ${parts[0]} on line ${pc + 1}`);
                }
                instruction.push(registerSet[parts[2]]);
            }

            // REG C
            if (["ADD","SUB","NOR","AND","XOR"].includes(parts[0])) {
                if (registerSet[parts[3]] === undefined) {
                    throw new Error(`Invalid reg C for ${parts[0]} on line ${pc + 1}`);
                }
                instruction.push(registerSet[parts[3]]);
            }

            // VALUE & ADDRESS
            if (["STR", "LOD", "LDR", "ADR"].includes(parts[0])) {
                const value = parseInt(parts[2]);
                if (value < 0 || value >= 256) {
                    throw new Error(`Invalid value for ${parts[2]} on line ${pc + 1}`);
                }
                instruction.push(value);
            }

            // LABEL / TODO: actually make this work
            if (["JMP", "CAL", "BRH"].includes(parts[0])) {
                console.log(labels[parts[1]]);
                if (labels[parts[1]] === undefined) {
                    throw new Error(`Unknown label: ${parts[1]} on line ${pc + 1}`);
                }
                instruction.push(labels[parts[1]]);
            }

            // CONDITION
            if (parts[0] === "BRH") {
                if (parseInt(parts[2]) !== (parseInt(parts[2]) % (2 ** 2))) {
                    throw new Error(`Invalid condition for BRH on line ${pc + 1}`);
                }
                instruction.push(parseInt(parts[2]));
            }
            
            instructions.push(instruction);
        }

        // pass five, instruction to binary
        for (const instruction of instructions) {
            const opcode = instruction[0]; 
            const operand = instruction.slice(1).reduce((acc, val, index) => {
                const shift = (index === 3 || index === 2) ? 4 : 8;
                return (acc << shift) | val;
            }, 0);
            const binary = ((opcode << 12) | operand).toString(2).padStart(16, '0');
            assembled = new Uint16Array([...assembled, parseInt(binary, 2)]);
        }

        return assembled;
    }
}

File: src/assembler.rs
use bitflags::bitflags;
use std::collections::HashMap;

bitflags! {
    struct InstructionSet: u8 {
        const HLT = 0b0000;
        const ADD = 0b0001;
        const SUB = 0b0010;

        const NOR = 0b0011;
        const AND = 0b0100;
        const XOR = 0b0101;

        const MOV = 0b0110;
        const LDR = 0b0111;
        const ADR = 0b1000;

        const JMP = 0b1001;
        const BRH = 0b1010;
        const CAL = 0b1011;

        const RET = 0b1100;
        const PGE = 0b1101;
        const LOD = 0b1110;

        const STR = 0b1111;
    }

    struct RegisterSet: u8 {
        const R0 = 0b0000;
        const R1 = 0b0001;
        const R2 = 0b0010;
        const R3 = 0b0011;
        const R4 = 0b0100;
        const R5 = 0b0101;
        const R6 = 0b0110;
        const R7 = 0b0111;
        const R8 = 0b1000;
        const R9 = 0b1001;
        const R10 = 0b1010;
        const R11 = 0b1011;
        const R12 = 0b1100;
        const R13 = 0b1101;
        const R14 = 0b1110;
        const R15 = 0b1111;
    }
}

pub struct Assembler;
impl Assembler {
    fn parse(code: String) -> Vec<Vec<String>> {
        code.lines()
            .filter(|l| !l.trim().is_empty() && !l.trim().starts_with(';'))
            .map(|l| l.trim().split_whitespace().map(String::from).collect())
            .collect()
    }

    pub fn assemble(code: String) -> Vec<u16> {
        let mut assembled: Vec<u16> = Vec::new();
        let mut definitions: HashMap<String, u8> = HashMap::new();
        let mut labels: HashMap<String, u8> = HashMap::new();
        let mut address: u8 = 0;

        let parsed = Self::parse(code);

        // first pass; collect keywords
        for parts in &parsed {
            if parts.len() == 3 && parts[0] == "define" {
                definitions.insert(parts[1].clone(), parts[2].parse::<u8>().unwrap());
                address += 1; // i think?!
            } else if parts.len() == 1 && parts[0].ends_with(':') {
                labels.insert(parts[0][..parts[0].len() - 1].to_string(), address);
            } else {
                address += 1;
            }
        }

        // second pass; assemble
        for parts in parsed {
            // skip keywords
            if parts.len() == 1 && parts[0].ends_with(':') {
                continue;
            }
            if parts[0] == "define" {
                continue;
            }

            let opcode = InstructionSet::from_name(parts[0].as_str()).expect("opcode should exist");
            let mut instruction = (opcode.bits() as u16) << 12;

            let get_register = |part: &String| RegisterSet::from_name(part.to_uppercase().as_str()).expect("register should exist").bits();
            match parts[0].as_str() {
                "HLT" | "RET" => {
                    assert_eq!(parts.len(), 1, "Invalid operand count");
                    instruction |= 0b_0000_0000_0000_0000;
                }
                "LDR" | "ADR" => {
                    assert_eq!(parts.len(), 3, "Invalid operand count");
                    let reg_a = get_register(&parts[1]);
                    let value = match definitions.get(parts[2].as_str()) {
                        Some(def) => *def,
                        None => parts[2].parse::<u8>().expect("Invalid immediate value"),
                    };
                    instruction |= (reg_a as u16) << 8;
                    instruction |= value as u16;
                }
                "JMP" | "CAL" | "BRH" => {
                    assert_eq!(parts.len(), 2, "Invalid operand count");
                    let address = labels.get(&parts[1]).copied().expect("Unknown label");
                    instruction |= address as u16;
                }
                _ => {
                    assert!(parts.len() >= 2, "Invalid operand count");
                    let reg_a = get_register(&parts[1]);
                    instruction |= (reg_a as u16) << 8;

                    if parts.len() > 2 {
                        let reg_b = get_register(&parts[2]);
                        instruction |= (reg_b as u16) << 4;
                    }

                    if parts.len() > 3 {
                        let reg_c = get_register(&parts[3]);
                        instruction |= reg_c as u16;
                    }
                }
            }

            assembled.push(instruction);
        }
        assembled
    }
}
